cpp/src/arrow/csv/parser.cc

```diff
@@ -183,52 +296,133 @@ Status BlockParser::ParseLine(const char* data, const char* data_end, bool is_fi
     return Status::OK();
   }
   // Truncated line at end of block, rewind parsed state
-  RewindState();
+  values_writer.RollbackLine();
+  parsed_writer.RollbackLine();
   return Status::OK();
 }
 
-Status BlockParser::DoParse(const char* start, uint32_t size, bool is_final,
-                            uint32_t* out_size) {
-  num_rows_ = 0;
-  // These don't shrink the allocated capacity, so reuses of BlockParser
-  // avoid most allocations when appending data
-  parsed_.clear();
-#ifdef CSV_PARSER_USE_BITFIELD
-  values_.clear();
-  values_.push_back({0, false});
-#else
-  offsets_.clear();
-  offsets_.push_back(0);
-  quoted_.clear();
-#endif
-
-  const char* data = start;
-  const char* data_end = start + size;
-
-  while (data < data_end && num_rows_ < max_num_rows_) {
+template <typename SpecializedOptions, typename ValuesWriter, typename ParsedWriter>
+Status BlockParser::ParseChunk(ValuesWriter& values_writer, ParsedWriter& parsed_writer,
```

Comment:
Consider using `T*` instead of `T&` for these first two arguments

Reply:
Will do.